includes/class-llms-media-protector.php

```diff
@@ -0,0 +1,640 @@
+	 *
+	 * Add `$_SERVER['MOD_X_SENDFILE_ENABLED'] = '1';` in `wp-config.php` if web server auto-detection isn't working.
+	 *
+	 * IIS administrators may want to use {@see https://github.com/stakach/IIS-X-Sendfile-plugin}.
+	 *
+	 * This method calls the {@see https://www.php.net/manual/function.exit.php `exit()`} function and does not return.
+	 *
+	 * @since [version]
+	 *
+	 * @param string $file_name The file path and name.
+	 * @param int    $media_id  The post ID of the media file.
+	 * @return void
+	 */
+	protected function send_file( string $file_name, int $media_id ) {
+
+		$server_software = $_SERVER['SERVER_SOFTWARE'];
+		$this->send_headers( $file_name, $media_id );
+
+		if (
+			'1' === $_SERVER['MOD_X_SENDFILE_ENABLED'] ||
+			( function_exists( 'apache_get_modules' ) && in_array( 'mod_xsendfile', apache_get_modules(), true ) ) ||
+			stristr( $server_software, 'cherokee' ) ||
+			stristr( $server_software, 'lighttpd' )
+		) {
+			header( "X-Sendfile: $file_name" );
+
+		} elseif ( stristr( $server_software, 'nginx' ) ) {
+			/**
+			 * @TODO Test NGINX.
+			 * @see https://www.nginx.com/resources/wiki/start/topics/examples/xsendfile/
+			 * @see https://woocommerce.com/document/digital-downloadable-product-handling/#nginx-setting
+			 */
+			// NGINX requires a URI without the server's root path.
+			$nginx_file_name = substr( $file_name, strlen( ABSPATH ) - 1 );
+			header( "X-Accel-Redirect: $nginx_file_name" );
+
+		} else {
+			$this->read_file( $file_name );
+		}
+
+		exit();
+	}
+
+	/**
+	 * Send headers for the download.
+	 *
+	 * @since [version]
+	 *
+	 * @param string $file_name The file path and name.
+	 * @param int    $media_id  The post ID of the media file.
+	 * @return void
+	 */
+	protected function send_headers( string $file_name, int $media_id ) {
+
+		$file_size = @filesize( $file_name ); // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged
+		if ( ! $file_size ) {
+			return;
+		}
+
+		$media_file   = get_post( $media_id );
+		$content_type = $media_file->post_mime_type;
+
+		header( "Content-Type: $content_type" );
+		header( "Content-Length: $file_size" );
+		header( 'X-Robots-Tag: noindex, nofollow', true );
+	}
+
+	/**
+	 * Sends a header that redirects the HTTP client to the media file's URL.
+	 *
+	 * This method calls the {@see https://www.php.net/manual/function.exit.php `exit()`} function and does not return.
+	 *
+	 * @since [version]
+	 *
+	 * @param int $media_id The post ID of the media file.
+	 * @return void
+	 */
+	protected function send_redirect( int $media_id ) {
+
+		$url = $this->get_media_url( $media_id );
+		header( "Location: $url" );
+		exit();
+	}
+
+	/**
+	 * Serves the requested media file to the HTTP client.
+	 *
+	 * This method calls the {@see https://www.php.net/manual/function.exit.php `exit()`} function and does not return.
+	 *
+	 * @todo Maybe respond with 404 Not Found?
+	 * @todo Maybe respond with 401 Unauthorized?
```

Comment:
The 403 response should respond with 401 if the user isn't logged in and 403 if the user is.

Reply:
[401 Unauthorized](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401) is supposed to be sent with a [WWW-Authenticate](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate) header to allow the user to authenticate in the little web browser pop up window. I'm guessing that if the user is not logged in, we should send them to the log in page instead?

Reply:
I'm going to say I wasn't aware of this but I guess I am since I wrote this https://github.com/gocodebox/lifterlms-rest/blob/9215df26a9a589d5a5488f019339dd83c1b3c15d/includes/class-llms-rest-authentication.php#L274-L291

I don't believe WordPress's default authentication layer is going to work properly with Basic Auth in this case. What I've just read (briefly) recommends using 403 in the kind of scenario we're in. I don't feel like that's the perfectly valid way to do it but I also read that performing a redirect isn't the proper way to do it either. So... /shrug

Reply:
After more research, it seems that 403 is the proper response code to use. 401 is meant to ask the user for credentials. We don't want to display a login page while displaying an element on the requested page. Basic authentication isn't normally used on WordPress sites.